pkg/cassandra/rack.go

```diff
@@ -0,0 +1,94 @@
+package cassandra
+
+import (
+	cassdcapi "github.com/k8ssandra/cass-operator/operator/pkg/apis/cassandra/v1beta1"
+	"github.com/k8ssandra/k8ssandra-operator/pkg/utils"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sort"
+)
+
+const zoneLabel = "failure-domain.beta.kubernetes.io/zone"
+
+func BuildRackAffinity(dc *cassdcapi.CassandraDatacenter, rackName string) *corev1.Affinity {
+	affinity := &corev1.Affinity{}
+	affinity.NodeAffinity = calculateRackAffinity(dc, rackName)
+	affinity.PodAntiAffinity = calculateRackAntiAffinity(dc)
+	return affinity
+}
+
+// calculateRackAffinity provides a way to decide where to schedule pods within a statefulset based on labels
+func calculateRackAffinity(dc *cassdcapi.CassandraDatacenter, rackName string) *corev1.NodeAffinity {
+	labels := rackAffinityLabels(dc, rackName)
+	if len(labels) == 0 {
+		return nil
+	}
+
+	var nodeSelectors []corev1.NodeSelectorRequirement
+
+	// we make a new map in order to sort because a map is random by design
+	keys := make([]string, 0, len(labels))
+	for key := range labels {
+		keys = append(keys, key)
+	}
+	sort.Strings(keys) // Keep labels in the same order across statefulsets
+	for _, key := range keys {
+		selector := corev1.NodeSelectorRequirement{
+			Key:      key,
+			Operator: corev1.NodeSelectorOpIn,
+			Values:   []string{labels[key]},
+		}
+		nodeSelectors = append(nodeSelectors, selector)
+	}
+
+	return &corev1.NodeAffinity{
+		RequiredDuringSchedulingIgnoredDuringExecution: &corev1.NodeSelector{
+			NodeSelectorTerms: []corev1.NodeSelectorTerm{{
+				MatchExpressions: nodeSelectors,
+			}},
+		},
+	}
+}
+
+func rackAffinityLabels(dc *cassdcapi.CassandraDatacenter, rackName string) map[string]string {
+	var labels map[string]string
+	racks := dc.GetRacks()
+	for _, rack := range racks {
+		if rack.Name == rackName {
+			labels = utils.MergeMap(rack.NodeAffinityLabels, dc.Spec.NodeAffinityLabels)
```

Comment:
Wanted to mention that this comes from cassdc api. As is, DC spec overrides rack spec. This seems in contradiction with K8ssandra's hierarchical principles but I didn't want to change the logic.

Reply:
That's a good point. Since it is the cluster controller who is creating the CassandraDatacenters, we can avoid using the `.Spec.NodeAffinityLabels` property. There is one caveat though. At some point we will need to adding an existing CassandraDatacenter to a K8ssandraCluster, particularly for migration scenarios. Maybe we should raise an issue in cass-operator for changing the behavior.